mls-mpm88-explained.cpp

```diff
@@ -0,0 +1,310 @@
+//  88-Line 2D Moving Least Squares Material Point Method (MLS-MPM) [Explained Version]
+
+// Uncomment this line for image exporting functionality
+#define TC_IMAGE_IO
+
+// Note: You DO NOT have to install taichi or taichi_mpm.
+// You only need [taichi.h] - see below for instructions.
+#include "taichi.h"
+
+using namespace taichi;
+using Vec = Vector2;
+using Mat = Matrix2;
+
+// Window
+const int window_size = 800;
+
+// Grid resolution (cells)
+const int n = 80;
+
+const real dt = 1e-4_f;
+const real frame_dt = 1e-3_f;
+const real dx = 1.0_f / n;
+const real inv_dx = 1.0_f / dx;
+
+// Snow material properties
+const auto particle_mass = 1.0_f;
+const auto vol = 1.0_f;        // Particle Volume
+const auto hardening = 10.0_f; // Snow hardening factor
+const auto E = 1e4_f;          // Young's Modulus
+const auto nu = 0.2_f;         // Poisson ratio
+const bool plastic = true;
+
+// Initial Lamé parameters
+const real mu_0 = E / (2 * (1 + nu));
+const real lambda_0 = E * nu / ((1+nu) * (1 - 2 * nu));
+
+struct Particle {
+  // Position and velocity
+  Vec x, v;
+  // Deformation gradient
+  Mat F;
+  // Cauchy-Green tensor
```

Comment:
Actually, is this an APIC thing? Just saw the comment below

Reply:
That's actually the APIC momentum. fixed in the updated version.